File: CodeForces/101-200/158/B/158B.cpp
// Problem 158B: Taxi
// https://codeforces.com/contest/158/problem/B
// Compiled with: g++ (Debian 12.2.0-14+deb12u1) 12.2.0
// Linux version: Linux kernel 6.1.0-10-amd64
// Submitted on: October 22nd, 2025
// Codeforces language used: C++23 (GCC 14-64, msys2)

#include <iostream>
#include <cmath>
using namespace std;

int main() {
    int n;
    cin >> n;

    // We don't need to keep track of fours separately here
    // Every group of 4 fills an entire taxi, so any groups of 4 add one taxi need
    int ones = 0, twos = 0, threes = 0, total_taxi = 0;

    for(int i = 0; i < n; i++) {
        int input;
        cin >> input;
        if (input == 1){
             ones++;
        } else if (input == 2) {
            twos++;
        } else if (input == 3) {
            threes++;
        } else {
            total_taxi++;
        }
    }

    // Every group of three must have its own taxi
    // A group of one can ride with them
    total_taxi += threes;
    // If there are more groups of one than three, ones will still be positive
    // Otherwise, set ones to zero.
    ones = max(ones-threes, 0);

    // There is no set of ones+twos that would require anything special
    // Assume all are groups of one for this purpose
    total_taxi += ceil(((ones+(twos*2.0))/4));

    cout << total_taxi << "\n";
}

File: CodeForces/101-200/158/B/158B.hs
-- Problem 158B: Taxi
-- https://codeforces.com/contest/158/problem/B
-- Executed with: The Glorious Glasgow Haskell Compilation System, version 9.6.6
-- Linux version: Linux kernel 6.17.0-6-generic
-- Submitted on: October 29th, 2025
-- Codeforces language used: Haskell GHC 8.10.1

main :: IO ()
main = do
    _ <- getLine
    line2 <- getLine
    let groups = map read (words line2) :: [Int]
    let threes = length (filter (==3) groups)
    let ones = max (length (filter (==1) groups) - threes) 0
    let twos = length (filter (==2) groups)
    let fours = length (filter (==4) groups)

    let extras = (ones + (twos*2) + 3) `div` 4

    let total = fours+threes+extras

    print total
